File: server/apps/nexotype/schemas/knowledge_graph/pathway_membership_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class PathwayMembershipCreate(BaseModel):
    """Schema for creating a new pathway membership"""
    protein_id: int
    pathway_id: int
    role: str | None = None


class PathwayMembershipUpdate(BaseModel):
    """Schema for updating a pathway membership"""
    protein_id: int = None
    pathway_id: int = None
    role: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class PathwayMembershipDetail(BaseModel):
    """Schema for pathway membership details"""
    id: int
    protein_id: int
    pathway_id: int
    role: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class PathwayMembershipListResponse(BaseModel):
    """Response schema for listing pathway memberships"""
    success: bool
    data: list[PathwayMembershipDetail] | None = None
    count: int = 0
    error: str | None = None


class PathwayMembershipResponse(BaseModel):
    """Response schema for single pathway membership operations"""
    success: bool
    data: PathwayMembershipDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/knowledge_graph/pathway-membership.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const PathwayMembershipSchema = z.object({
  id: z.number(),
  protein_id: z.number().int(),
  pathway_id: z.number().int(),
  role: z.string().max(100).nullable(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreatePathwayMembershipSchema = z.object({
  protein_id: z.number().int(),
  pathway_id: z.number().int(),
  role: z.string().max(100).nullable().optional(),
});

export const UpdatePathwayMembershipSchema = z.object({
  protein_id: z.number().int().optional(),
  pathway_id: z.number().int().optional(),
  role: z.string().max(100).nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type PathwayMembership = z.infer<typeof PathwayMembershipSchema>;
export type CreatePathwayMembership = z.infer<typeof CreatePathwayMembershipSchema>;
export type UpdatePathwayMembership = z.infer<typeof UpdatePathwayMembershipSchema>;

// ==========================================
// Response Types
// ==========================================

export type PathwayMembershipResponse = {
  success: boolean;
  data?: PathwayMembership;
  error?: string;
};

export type PathwayMembershipsResponse = {
  success: boolean;
  data?: PathwayMembership[];
  count?: number;
  error?: string;
};
